packages/react-aria-components/docs/Checkbox.mdx

```diff
@@ -360,6 +360,60 @@ Checkbox supports the `name` and `value` props for integration with HTML forms.
+
+<ContextTable components={['Checkbox']} docs={docs} />
+
+This example shows a `CheckboxDescription` component that accepts a checkbox in its children and renders a description element below it. It uses the [useId](useId.html) hook to generate a unique id for the description, and associates it with the checkbox via the `aria-describedby` attribute passed to the `CheckboxContext` provider.
+
+```tsx example
+import {CheckboxContext} from 'react-aria-components';
+import {useId} from 'react-aria';
+
+interface CheckboxDescriptionProps {
+  children?: React.ReactNode,
+  description?: string
+}
+
+function CheckboxDescription({children, description}: CheckboxDescriptionProps) {
+  let descriptionId = useId();
+  return (
+    <div>
+      {/*- begin highlight -*/}
+      <CheckboxContext.Provider value={{'aria-describedby': descriptionId}}>
+      {/*- end highlight -*/}
+        {children}
+      </CheckboxContext.Provider>
+      <small id={descriptionId}>{description}</small>
+    </div>
+  );
+}
+
+<CheckboxDescription description="You will receive our newsletter once per week. Unsubscribe at any time.">
+  <MyCheckbox defaultSelected>Subscribe</MyCheckbox>
+</CheckboxDescription>
+```
+
 ### Hooks
 
-If you need to customize things further, such as accessing internal state or customizing DOM structure, you can drop down to the lower level Hook-based API. See [useCheckbox](useCheckbox.html) for more details.
+If you need to customize things further, such as intercepting events or customizing DOM structure, you can drop down to the lower level Hook-based API. Consume from `CheckboxContext` in your component with <TypeLink links={docs.links} type={docs.exports.useContextProps} /> to make it compatible with other React Aria Components. See [useCheckbox](useCheckbox.html) for more details.
+
+```tsx
+import type {CheckboxProps} from 'react-aria-components';
+import {CheckboxContext, useContextProps} from 'react-aria-components';
+import {useToggleState} from 'react-stately';
+import {useCheckbox} from 'react-aria';
+
+const MyCheckbox = React.forwardRef((props: CheckboxProps, ref: React.ForwardedRef<HTMLInputElement>) => {
+  // Merge the local props and ref with the ones provided via context.
+  ///- begin highlight -///
+  [props, ref] = useContextProps(props, ref, CheckboxContext);
+  ///- end highlight -///
+
+  let state = useToggleState(props);
+  let {inputProps} = useCheckbox(props, state, ref);
+  return <input {...inputProps} ref={ref} />;
```

Comment:
I debated including this example at all since it doesn't really do anything that a RAC checkbox couldn't do - it's not a real example of when you'd actually need to drop down to hooks. I didn't really do this for any of the other components. I guess I left it because checkbox is used within lots of other components (e.g. table/gridlist), so showing how to consume the contexts is important. Should I do this on all of the other pages that don't have an example? Some of them are a bit harder to build though so the code would get long...

Reply:
I agree, this is a bit lackluster, I'll keep trying to think of a different example. If we don't include it on other pages, maybe we can link to it as a concept, or maybe we just have a page for Context (name tbd) which describes this and all pages link to it?

Reply:
I just noticed the button one has a motion example, maybe that'd be better here?

Reply:
+1 to having a separate page detailing the aria hooks + useContextProps approach here. Introducing the concept + having a single/couple examples I feel should be sufficient for end users IMO

Reply:
Yeah I'd like to have an "Advanced customization" overview page that covers contexts, slots, hooks, etc. in general. But I also know that most people won't read that and will go straight to the component they are interested in, hence all the duplicated content on every page. It also might be useful to show what specific context to consume from for each specific component, along with the TS types for the props and ref that you need. Maybe we can show the setup but omit the actual implementation and just show a `// ...`? idk

Reply:
Just to note, I'm not saying to do this in only an advanced customization page, I think it's fine to have it duplicated on every component for the reason you mentioned. I was just thinking it'd be good if each of those also linked to a place where we do a deep dive and show some of the more compelling cases for using it. I think that'd help vs showing people something they can already do with RAC. Otherwise we're going to see people doing this Checkbox as it's shown here instead of just using RAC, and that seems a shame.

Instead, we could make the checkbox section short, and link to the combined advanced customization page where we can show a more complex multi-component use case that would be closer to real life.

Reply:
>  But I also know that most people won't read that and will go straight to the component they are interested in, hence all the duplicated content on every page.

Fair point, I think linking to a "Advanced customization" page like Rob suggested above and just detailing the available contexts + TS types for props/ref like you mentioned should be sufficient for components that don't have a good standalone "drop down to hooks" example.